### PR TITLE
refactor(tauri.js) rewrite API module in TypeScript, closes #679 #435

### LINKED ISSUE #679: API rework - TypeScript edition
**Is your feature request related to a problem? Please describe.**
Improve [the current API](https://github.com/tauri-apps/tauri/tree/dev/cli/tauri.js/api) so we have typings and ES5 code.

**Describe the solution you'd like**
Change .js for .ts. The output code should keep the current directory/file structure.

Change current JSDoc types for TS types:
- `HttpOptions` in http.js (should match again the [Rust definition](https://github.com/tauri-apps/tauri/blob/dev/tauri-api/src/http.rs));
- `EventCallback` in event.js.

**Additional context**
Some more context about the Rust code under the hood:
https://github.com/tauri-apps/tauri/blob/406dea79ed7f187e773078761c787ebe6811e97e/tauri/src/endpoints.rs

### PR DESCRIPTION
**Pull request by nothingismagick**
This PR moves `/cli/tauri.js/api` to `/cli/tauri.js/api-src`, adds rollup transpilation to give us some nice shiny new assets:

- UMD (monolith, minified)
- CJS (minified)
- ESM (marked as pure)

Closes  #679 .

Todo:
- [x] Update GH Actions
- [ ] Make sure the types are proper and available
- [ ] Consider a method for treeshaking the UMD

**What kind of change does this PR introduce?** (check at least one)

- [ ] Bugfix
- [x] Feature
- [ ] New Binding Issue #___
- [ ] Code style update
- [x] Refactor
- [x] Build-related changes
- [ ] Other, please describe:

**Does this PR introduce a breaking change?**

I assume this will break the API for some consumers.

**The PR fulfills these requirements:**

- [x] It's submitted to the `dev` branch and _not_ the `master` branch
- [x] When resolving a specific issue, it's referenced in the PR's title (e.g. `fix: #xxx[,#xxx]`, where "xxx" is the issue number)
- [x] A change file is added if any packages will require a version bump due to this PR per [the instructions in the readme](https://github.com/tauri-apps/tauri/blob/dev/.changes/readme.md).

If adding a **new feature**, the PR's description includes:
- [x] A convincing reason for adding this feature (to avoid wasting your time, it's best to open a suggestion issue first and wait for approval before working on it)

**Other information:**

### CODE REVIEW

**Comment by nothingismagick** [comment]
@jbolda - not sure how to best make sure `yarn build:api` gets run in the `release-npm.yml` workflow.

**Comment by lucasfernog** [comment]
@nothingismagick maybe just add to the `build-release` script on package.json?

**Comment by jbolda** [comment]
Or perhaps add it to the prepublishOnly script?

File: cli/tauri.js/api-src/cli.ts [hunk 1]
```diff
@@ -3,7 +3,7 @@ import tauri from './tauri'
 /**
  * gets the CLI matches
  */
-function getMatches() {
+function getMatches(): any {
```
Comment (lucasfernog) [inline cli/tauri.js/api-src/cli.ts:6]: I'll add the typedef here

File: cli/tauri.js/tsconfig.json [hunk 2]
```diff
@@ -12,7 +12,7 @@
     "paths": {
       "types": ["src/types"]
     },
-    "resolveJsonModule": true
```
Comment (lucasfernog) [inline cli/tauri.js/tsconfig.json:15]: why was this removed?

File: cli/tauri.js/tsconfig.json [hunk 3]
```diff
@@ -12,7 +12,7 @@
...
   },
-  "include": ["src"]
+  "include": ["src"],
+  "// extends": "./api/tsconfig.json"
```
Comment (lucasfernog) [inline cli/tauri.js/tsconfig.json:17]: maybe this one should be removed

File: maskfile.md [hunk 4]
```diff
@@ -16,7 +16,7 @@ cargo install --path cli/tauri-bundler --force
 cargo install cargo-web 			# used by example rust/yew
 
 cd cli/tauri.js
-yarn && yarn build
+yarn && yarn build && yarn rollup -c
```
Comment (lucasfernog) [inline maskfile.md:19]: maybe `yarn build` should do everything?

File: cli/tauri.js/tsconfig.json (hunk 3 above)
Comment (nothingismagick) [reply cli/tauri.js/tsconfig.json:17]: yeah - i missed that in my review.

File: cli/tauri.js/tsconfig.json (hunk 2 above)
Comment (nothingismagick) [reply cli/tauri.js/tsconfig.json:15]: its on line 9

**Comment by nothingismagick** [comment]
Something is up with the FS builtin paths (in the example)
image
